src/main/java/com/google/devtools/build/lib/rules/proto/ProtoCompileActionBuilder.java

```diff
@@ -288,21 +280,16 @@ private static String getOutputDirectory(RuleContext ruleContext) {
       result.add("--disallow_services");
     }
     if (checkStrictImportPublic) {
-      NestedSet<Pair<Artifact, String>> protosInExports =
-          protoInfo.getExportedProtoSourcesImportPaths();
-      if (protosInExports.isEmpty()) {
+      NestedSet<ProtoSource> publicImportSources = protoInfo.getPublicImportSources();
+      if (publicImportSources.isEmpty()) {
         // This line is necessary to trigger the check.
         result.add("--allowed_public_imports=");
       } else {
         result.addAll(
             "--allowed_public_imports",
             VectorArg.join(":")
-                .each(protosInExports)
-                .mapped(
-                    new ExpandToPathFnWithImports(
-                        getOutputDirectory(ruleContext),
-                        protoInfo.getTransitiveProtoSourceRoots(),
-                        siblingRepositoryLayout)));
+                .each(publicImportSources)
+                .mapped((s, args) -> s.getImportPath().getPathString()));
```

Comment:
Mind extracting these to a separate class just like the other argument map functions? (one of our internal tests requires that, long story)